DOCSP-25962: MONGODB-AWS aws sdk info (#298)
mongodb/docs-java: source/fundamentals/auth.txt

``MONGODB-AWS``
~~~~~~~~~~~~~~~

..
   The MONGODB-AWS section structure was updated for v4.8 of the driver. Avoid
   backporting any documentation that might not apply to a prior version.

.. note::

   The MONGODB-AWS authentication mechanism is available for MongoDB
   deployments on MongoDB Atlas.

The ``MONGODB-AWS`` authentication mechanism uses your Amazon Web Services
Identity and Access Management (AWS IAM) credentials to authenticate your
user. To learn more about configuring MongoDB Atlas, see the
:atlas:`Set Up Passwordless Authentication with AWS IAM Roles </security/passwordless-authentication/#set-up-passwordless-authentication-with-aws-iam-roles>`
guide.

You can authenticate using the ``MONGODB-AWS`` mechanism by providing
your AWS IAM credentials in one of the following ways:

> - As environment variables

Reviewer: I'm still confused by this, based on the shared doc we are now working off of.

Author: I see, did I conflate "From ENVIRONMENT" with "environment variables"? If so, what else is from the environment that I missed that isn't the environment variables?
If not, doesn't the current organization correspond directly with "FROM ENVIRONMENT" and "FROM MongoClient CONFIGURATION"? There are only two headers within the section now which I intended to correspond to this bulleted list and the doc:
"Specify Your Credentials in Environment Variables"
"Specify Your Credentials in MongoClient Configuration"

Reviewer: > Did I conflate "From ENVIRONMENT" with "environment variables"? If so, what else is from the environment that I missed that isn't the environment variables?

Yes, exactly.  The environment also includes the HTTP endpoints for EC2 and ECS

Author: Trying to figure out how to define "from environment" to a user -- what are the boundaries of that term? I can't think of any specific term that might express environment variables (for secret, tokens, ECS endpoint) and lack of environment variables (EC2 endpoint) which is what I think we intend to communicate.

Would "As environment variables or by using the AWS SDK for Java" be more appropriate?

Reviewer: No, because the SDK also, and first, looks for the presence of env vars.

- In a ``MongoClient`` configuration

The next sections demonstrate how to specify your credentials each way
and use the following placeholders:

* ``awsKeyId`` - value of your AWS access key ID
* ``awsSecretKey`` - value of your AWS secret access key
* ``atlasUri`` - network address of your MongoDB Atlas deployment
* ``hostname`` - hostname of your MongoDB Atlas deployment
* ``port`` - port of your MongoDB Atlas deployment
* ``awsSessionToken`` - value of your AWS session token *(optional)*

.. _java-mongodb-aws-env-variables:

Specify Your Credentials in Environment Variables
++++++++++++++++++++++++++++++++++++++++++++++++++

You can provide your AWS IAM credentials by instructing the driver to
use the ``MONGODB-AWS`` authentication mechanism and by setting the
appropriate environment variables.

.. tip:: Add the AWS SDK for Java

>    When you use environment variables to specify your AWS IAM credentials, we

Reviewer: This is not completely accurate.  We also recommend this for EC2, where there are no environment variables set, and for ECS, which only has the single `AWS_CONTAINER_CREDENTIALS_RELATIVE_URI` environment variable

   recommend adding the AWS SDK for Java (v1 or v2) dependency to your
   project. In addition to authentication, the SDK offers the following
   features:

   - Credential caching which helps your application avoid rate limiting
   - Credential provider management for use with the `Elastic Kubernetes Service <https://docs.aws.amazon.com/eks/latest/userguide/what-is-eks.html>`__.

>    To use the AWS SDK, include the dependencies for it in your project and set

Reviewer: Consider including the actual coordinates of the dependencies

   * `software.amazon.awssdk:auth:<latest v2 version>` (driver currently tests with 2.18.9)
   * `com.amazonaws:aws-java-sdk-core:<latest v1 version>` (driver currently tests with 1.12.337)

   the credentials for one of the authentication methods in the provider
   chain.

>    If your project includes AWS SDK for Java **v2**, the driver passes the

Reviewer: > the driver passes the credentials

The driver passes nothing.  The credentials are obtained by the SDK from the environment

   credentials to the `DefaultCredentialsProvider <https://sdk.amazonaws.com/java/api/latest/software/amazon/awssdk/auth/credentials/DefaultCredentialsProvider.html>`__.

   If your project includes the AWS SDK for Java **v1** and does not include

>    **v2** SDK, the driver passes the credentials to the `DefaultAWSCredentialsProviderChain <https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/auth/DefaultAWSCredentialsProviderChain.html>`__.

Reviewer: As above

   To learn more about the SDK, see the `AWS SDK for Java <https://docs.aws.amazon.com/sdk-for-java/index.html>`__
   AWS documentation.

Use one of the following methods to set the authentication mechanism:

- Call the ``MongoCredential.createAwsCredential()`` method to construct a
  ``MongoClient`` as shown in the following example:

  .. literalinclude:: /includes/fundamentals/code-snippets/MongoDbAwsAuth.java
     :language: java
     :dedent:
     :start-after: start mechOnlyMongoCredential
     :end-before: end mechOnlyMongoCredential
     :emphasize-lines: 1,7

- Add the mechanism as a parameter to the connection string that you pass to
  instantiate the ``MongoClient`` as shown in the following example:

  .. literalinclude:: /includes/fundamentals/code-snippets/MongoDbAwsAuth.java
     :language: java
     :dedent:
     :start-after: start mechOnlyConnectionString
     :end-before: end mechOnlyConnectionString

The following example shows how you can define your AWS IAM credentials as
environment variables by using ``bash`` or a similar shell:

.. code-block:: bash

   export AWS_ACCESS_KEY_ID=<awsKeyId>
   export AWS_SECRET_ACCESS_KEY=<awsSecretKey>

>    export AWS_SECRET_TOKEN=<awsSessionToken>

Reviewer: AWS_SECRET_TOKEN. -> AWS_SESSION_TOKEN

> Omit the line containing ``AWS_SECRET_TOKEN`` if you don't need an AWS

Reviewer: AWS_SECRET_TOKEN -> AWS_SESSION_TOKEN

session token for that role.

To authenticate by using **ECS container credentials**, specify the relative
URI of the container in an environment variable called
``AWS_CONTAINER_CREDENTIALS_RELATIVE_URI``.

To authenticate using **EC2 container credentials**, make sure none of the
aforementioned environment variables are set.

.. _java-mongodb-aws-mongoclient-configuration:

Specify Your Credentials in MongoClient Configuration
+++++++++++++++++++++++++++++++++++++++++++++++++++++

You can configure a ``MongoClient`` instance to use the ``MONGODB-AWS``
authentication mechanism by passing your AWS IAM credentials to the
`createAwsCredential() <{+api+}/apidocs/mongodb-driver-core/com/mongodb/MongoCredential.html#createAwsCredential(java.lang.String,char%5B%5D)>`__
method.

.. important::

   If you configure your ``MongoClient`` with this method, the

>    driver does not check for AWS IAM credentials in environment variables.

Reviewer: in environment variables or EC2/ECS endpoints

.. note::

   The MONGODB-AWS authentication mechanism is available for MongoDB
   deployments on MongoDB Atlas.

The ``MONGODB-AWS`` authentication mechanism uses your Amazon Web Services
Identity and Access Management (AWS IAM) credentials to authenticate your
user. To learn more about configuring MongoDB Atlas, see the
:atlas:`Set Up Passwordless Authentication with AWS IAM Roles </security/passwordless-authentication/#set-up-passwordless-authentication-with-aws-iam-roles>`
guide.

You can authenticate using the ``MONGODB-AWS`` mechanism by providing
your AWS IAM credentials in one of the following ways:

1.  Use the AWS SDK (v1 or v2) to authenticate an IAM role.
#. Specify your AWS credentials in a ``MongoCredential`` instance or connection
   string.
#. Specify your AWS credentials in environment variables.
#. Specify an ECS/EC2 endpoint that hosts your container credentials.

.. important::

   The driver uses the preceding order to check for your credentials and uses
   the first method that it finds.

To learn how to refresh your AWS IAM credentials, see the
:ref:`<java-mongodb-aws-refresh-credentials>` section.

.. _java-mongodb-aws-sdk:

> Authenticate using the AWS SDK

Reviewer 2: Issue: Per [style guide](https://www.mongodb.com/docs/meta/style-guide/style/titles-and-headings/#guidelines-for-headline-style-capitalization), I think "using" should be capitalized.

Suggestion:

```suggestion
Authenticate Using the AWS SDK
```

+++++++++++++++++++++++++++++++

> You can authenticate with your AWS IAM role using the `AWS SDK for Java <https://docs.aws.amazon.com/sdk-for-java/index.html>`__

Reviewer 2: Issue: Per [style guide](https://www.mongodb.com/docs/meta/style-guide/terminology/alphabetical-terms/#term-using), "by using" would be clearer than "using" in this case. Otherwise it could mean that the AWS IAM role is using the AWS SDK for Java.

Suggestion:

```suggestion
You can authenticate with your AWS IAM role by using the `AWS SDK for Java <https://docs.aws.amazon.com/sdk-for-java/index.html>`__
```

versions 1 or 2 by including it in your project and authenticating with
the SDK..

If your project includes AWS SDK for Java **v1**, the driver uses the `DefaultAWSCredentialsProviderChain <https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/auth/DefaultAWSCredentialsProviderChain.html>`__
class to specify the credentials.

If your project includes AWS SDK for Java **v2**, the driver uses the `DefaultCredentialsProvider <https://sdk.amazonaws.com/java/api/latest/software/amazon/awssdk/auth/credentials/DefaultCredentialsProvider.html>`__
class to specify the credentials.

> Authenticate using AWS IAM Credentials

Reviewer 2: Issue: Per [style guide](https://www.mongodb.com/docs/meta/style-guide/style/titles-and-headings/#guidelines-for-headline-style-capitalization), I think "using" should be capitalized.

Suggestion:

```suggestion
Authenticate Using AWS IAM Credentials
```

++++++++++++++++++++++++++++++++++++++

You authenticate with your AWS IAM role by specifying your credentials in
a connection string, a ``MongoCredential`` instance, or in environment
variables.

The examples in this section use the following placeholders:

* ``awsKeyId`` - value of your AWS access key ID
* ``awsSecretKey`` - value of your AWS secret access key
* ``atlasUri`` - network address of your MongoDB Atlas deployment
* ``hostname`` - hostname of your MongoDB Atlas deployment
* ``port`` - port of your MongoDB Atlas deployment
* ``awsSessionToken`` - value of your AWS session token *(optional)*

.. _mongodb-aws-connection-string-example:

Connection String Example
^^^^^^^^^^^^^^^^^^^^^^^^^

To authenticate using a connection string, assign the ``authMechanism``
parameter the value ``"MONGODB-AWS"`` and replace the placeholders with your

> credentials in your connection string. The following code example shows how
> you can instantiate a ``MongoClient`` using your connection string:

Reviewer 2: Issue: Per [style guide](https://www.mongodb.com/docs/meta/style-guide/terminology/alphabetical-terms/#term-using), "using" isn't optimal here (the MongoClient could be using your connection string--which, yes, it kind of is, but I don't think that's what we're trying to convey here).

Suggestion:

```suggestion
credentials in your connection string. The following code example shows how
you can use your connection string to instantiate a ``MongoClient``:
```

Reviewer 2: There might be a more accurate and/or concise way to phrase this, but your revised version is unambiguous and understandable, so no changes required here.

.. literalinclude:: /includes/fundamentals/code-snippets/MongoDbAwsAuth.java
   :language: java
   :dedent:
   :start-after: start connectionString
   :end-before: end connectionString

If you need to specify an AWS session token, include it in the

> ``authMechanismProperties`` parameter using the format

Reviewer 2: Issue: See previous comments about "using"

```suggestion
``authMechanismProperties`` parameter in the format
```

Author: Thanks for spotting this. I didn't review the grammar of the existing content since that felt beyond the scope of what I was trying to accomplish, but could be difficult for the reviewer to identify. Will update to use "in".

``AWS_SESSION_TOKEN:<awsSessionToken>``. The following code example shows

> how you can instantiate a ``MongoClient`` that includes an AWS session token
> using your connection string:

Reviewer 2: Issue: See previous comments about "using"

Suggestion:

```suggestion
how you can use your connection string to instantiate a ``MongoClient`` that includes an AWS session token:
```

Author: Addressing this using the same solution I come up with for [this comment thread](https://github.com/mongodb/docs-java/pull/298#discussion_r1024210194).

Reviewer 2: Same feedback as previous thread.

.. note::

   The MONGODB-AWS authentication mechanism is available for MongoDB
   deployments on MongoDB Atlas.

The ``MONGODB-AWS`` authentication mechanism uses your Amazon Web Services
Identity and Access Management (AWS IAM) credentials to authenticate your
user. To learn more about configuring MongoDB Atlas, see the
:atlas:`Set Up Passwordless Authentication with AWS IAM Roles </security/passwordless-authentication/#set-up-passwordless-authentication-with-aws-iam-roles>`
guide.

You can authenticate using the ``MONGODB-AWS`` mechanism by providing
your AWS IAM credentials in one of the following ways:

> 1.  Use the AWS SDK (v1 or v2) to authenticate an IAM role.

Author: Looks like this comment is from an outdated version.

#. Specify your AWS credentials in a ``MongoCredential`` instance or connection
   string.
#. Specify your AWS credentials in environment variables.

> #. Specify an ECS/EC2 endpoint that hosts your container credentials.

Author: Looks like this comment is from an outdated version.

.. important::

   The driver uses the preceding order to check for your credentials and uses
   the first method that it finds.

To learn how to refresh your AWS IAM credentials, see the
:ref:`<java-mongodb-aws-refresh-credentials>` section.

.. _java-mongodb-aws-sdk:

Authenticate Using the AWS SDK
+++++++++++++++++++++++++++++++

You can use the `AWS SDK for Java <https://docs.aws.amazon.com/sdk-for-java/index.html>`__

> versions 1 or 2 to authenticate. You must include the AWS SDK in your project

Author: Looks like this comment is from an outdated version. Would "dependencies for it" be clear enough, or does it need to read "include a dependency on the AWS SDK"?

> versions 1 or 2 to authenticate. You must include the AWS SDK in your project

Reviewer: The latter is good.  And as I said in another comment, we could even provide the Maven coordinates, like we do with other optional dependencies

> and call the appropriate authentication methods.

Author: Would  "set the credentials for one of the methods in the provider chain" work here?

If your project includes AWS SDK for Java **v1**, the driver uses the `DefaultAWSCredentialsProviderChain <https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/auth/DefaultAWSCredentialsProviderChain.html>`__
class to specify the credentials.

> If your project includes AWS SDK for Java **v2**, the driver uses the `DefaultCredentialsProvider <https://sdk.amazonaws.com/java/api/latest/software/amazon/awssdk/auth/credentials/DefaultCredentialsProvider.html>`__

Author: Looks like this comment is from an outdated version.

  string as shown in the following example:

  .. code-block:: none

     mongodb://<atlasUri>/?authMechanism=MONGODB-AWS

- Set your AWS IAM role credentials in environment variables in your
  client execution environment. The following example shows how you can set
  them in a bash or similar shell:

  .. code-block:: bash

     export AWS_ACCESS_KEY_ID=<awsKeyId>
     export AWS_SECRET_ACCESS_KEY=<awsSecretKey>
     export AWS_SECRET_TOKEN=<awsSessionToken>

  You can omit the line containing ``AWS_SECRET_TOKEN`` if you do not use
  an AWS session token for that role.

AWS ECS/EC2 Endpoint Example
^^^^^^^^^^^^^^^^^^^^^^^^^^^^

To authenticate, specify the ECS or EC2 container relative URI in an
environment variable called ``AWS_CONTAINER_CREDENTIALS_RELATIVE_URI`` and
assign the value of your container relative URI to it.

When using bash or a similar shell in your client execution environment, you
can set the environment variable, replacing the value of the placeholder
with your relative URI:

.. code-block:: bash

   export AWS_CONTAINER_CREDENTIALS_RELATIVE_URI=<your ECS/EC2 endpoint>

.. important::

   To authenticate using an ECS/EC2 endpoint, you must unset the following

>    environment variables because the driver otherwise attempts to authenticate
>    using them:

Reviewer 2: Optional:

```suggestion
   environment variables because the driver otherwise attempts to use them to authenticate:
```

   - ``AWS_ACCESS_KEY_ID``
   - ``AWS_SECRET_ACCESS_KEY``
   - ``AWS_SESSION_TOKEN``

To learn more about the AWS container credential provider and container
credentials, see the AWS documentation on `Container credentials <https://docs.aws.amazon.com/sdkref/latest/guide/feature-container-credentials.html>`__.

.. _java-mongodb-aws-refresh-credentials:

> Refresh your Credentials

Reviewer 2: Issue: Per [style guide](https://www.mongodb.com/docs/meta/style-guide/style/titles-and-headings/#guidelines-for-headline-style-capitalization), I think "your" should be capitalized.

Suggestion:

```suggestion
Refresh Your Credentials
```

++++++++++++++++++++++++

The driver supports refreshing credentials to support assuming roles

> or using the `Elastic Kubernetes Service <https://docs.aws.amazon.com/eks/latest/userguide/what-is-eks.html>`__.

Reviewer 2: Issue: A lot of gerunds ("-ing" verbs) make this sentence a little confusing.

Suggestion: I'm not sure I have the technical knowledge to rephrase this accurately, unfortunately! Anything that reduces the number of gerunds should work.

Author: Thanks for spotting this. I didn't review the grammar of the existing content since that felt beyond the scope of what I was trying to accomplish, but could be difficult for the reviewer to identify. Will think of a better way to phrase it.

The following code example shows how you can specify a ``Supplier`` instance

> that refreshes your credentials.

Reviewer 2: Question: Did you intentionally use a period to introduce the following code sample? Would like to know why, if so. Most or all of the examples on this page use colons.
